packages/java-parser/src/comments.js

```diff
@@ -155,6 +155,61 @@ function shouldIgnore(node, comments, ignoredNodes) {
   }
 }
 
+function filterFormatterOffOn(comments) {
+  return [...comments].filter(comment =>
+    comment.image.match(
+      /(\/\/(\s*)@formatter:off(\s*))|(\/\*(\s*)@formatter:off(\s*)\*\/)|(\/\/(\s*)@formatter:on(\s*))|(\/\*(\s*)@formatter:on(\s*)\*\/)/gm
+    )
+  );
+}
+
+function matchFormatterOffOnPair(comments) {
+  let isPreviousCommentOff = false;
+  let isCurrentCommentOff = true;
+  const pairs = [];
+  let paired = {};
+  comments.forEach((comment, index, array) => {
+    isCurrentCommentOff =
+      comment.image.slice(comment.image.length - 3) === "off";
+
+    if (!isPreviousCommentOff) {
+      if (isCurrentCommentOff) {
+        paired.off = comment;
+      }
+    } else {
+      if (!isCurrentCommentOff) {
+        paired.on = comment;
+        pairs.push(paired);
+        paired = {};
+      }
+    }
+
+    if (index === array.length - 1 && isCurrentCommentOff) {
+      paired.on = undefined;
+      pairs.push(paired);
+      paired = {};
+    }
+
+    isPreviousCommentOff = isCurrentCommentOff;
+  });
+
+  return pairs;
+}
+
+function shouldNotFormat(node, commentPairs) {
+  const matchingPair = _.find(
+    commentPairs.reverse(),
```

Comment:
Be aware that this method is called for each node of the file, so the reverse method is called a lot. You can maybe use `findLast` from lodash to avoid calling it, or reverse it directly in the `setOffOnComments` to do it once

Reply:
I agree, you're right!